src/core/thread/network_data_leader.cpp

```diff
@@ -577,20 +575,200 @@ void Leader::SendServerDataResponse(const Coap::Header &aRequestHeader, const Ip
+                    hasRoute = FindHasRoute(*prefix);
+                    for (int i = 0; i < hasRoute->GetNumEntries(); i++)
+                    {
+                        hasRouteEntry = hasRoute->GetEntry(i);
+                        if (hasRouteEntry->GetRloc() == aRloc16)
+                        {
+                            *aIn = true;
+                            if (hasRoute->IsStable())
+                            {
+                                *aStable = true;
+                            }
+                        }
+                    }
+                    break;
+                default:
+                    break;
+                }
+
+                if ((*aIn) && (*aStable))
+                {
+                    ExitNow();
+                }
+
+                subCur = subCur->GetNext();
+            }
+        }
+        cur = cur->GetNext();
+    }
+
+exit:
+    return;
+}
+
+bool Leader::IsStableUpdated(uint16_t aRloc16, uint8_t *aTlvs, uint8_t aTlvsLength, uint8_t *aTlvsBase, uint8_t aTlvsBaseLength)
+{
+    bool rval = false;
+    NetworkDataTlv *cur = reinterpret_cast<NetworkDataTlv *>(aTlvs);
+    NetworkDataTlv *end = reinterpret_cast<NetworkDataTlv *>(aTlvs + aTlvsLength);
+    PrefixTlv *prefix;
+    PrefixTlv *prefixBase;
+    BorderRouterTlv *borderRouter;
+    HasRouteTlv *hasRoute;
+
+    while (1)
+    {
+        if (cur >= end)
+        {
+            ExitNow(rval = false);
+        }
+
+        if (cur->GetType() == NetworkDataTlv::kTypePrefix)
+        {
+            prefix = reinterpret_cast<PrefixTlv *>(cur);
+            if (cur->IsStable())
+            {
+                prefixBase = FindPrefix(prefix->GetPrefix(), prefix->GetPrefixLength(), aTlvsBase, aTlvsBaseLength);
+                if (!prefixBase)
+                {
+                    ExitNow(rval = true);
+                }
+
+                borderRouter = FindBorderRouter(*prefix);
+                if (borderRouter && (memcmp(borderRouter, FindBorderRouter(*prefixBase), borderRouter->GetLength()) != 0))
+                {
+                    ExitNow(rval = true);
+                }
+
+                hasRoute = FindHasRoute(*prefix);
+                if (hasRoute && (memcmp(hasRoute, FindHasRoute(*prefixBase), hasRoute->GetLength()) != 0))
+                {
+                    ExitNow(rval = true);
+                }
+            }
+        }
+        cur = cur->GetNext();
+   }
+
+exit:
+    return rval;
+}
+
 ThreadError Leader::RegisterNetworkData(uint16_t aRloc16, uint8_t *aTlvs, uint8_t aTlvsLength)
 {
     ThreadError error = kThreadError_None;
 
-    SuccessOrExit(error = RemoveRloc(aRloc16));
-    SuccessOrExit(error = AddNetworkData(aTlvs, aTlvsLength));
+    bool rlocIn = false;
+    bool rlocStable = false;
+    bool stableUpdated = false;
 
-    mVersion++;
-    mStableVersion++;
+    RlocLookup(aRloc16, &rlocIn, &rlocStable, mTlvs, mLength);
+
+    if (aTlvs == NULL)
+    {
+        if (rlocIn)
```

Comment:
Can we move this code block up into `RemoveBorderRouter` (plus the few extra lines at the beginning and end)?  It's a little confusing to have `RegisterNetworkData` do two different things based on whether or not `aTlvs` is `NULL`.